content/graph/HLD.h

```diff
@@ -1,112 +1,60 @@
  * Description: Decomposes a tree into vertex disjoint heavy paths and light
- *  edges such that the path from any leaf to the root contains at most log(n)
- *  light edges. The function of the HLD can be changed by modifying T, LOW and
- *  f. f is assumed to be associative and commutative.
- * Usage:
- *  HLD hld(G);
- *  hld.update(index, value);
- *  tie(value, lca) = hld.query(n1, n2);
- * Status: Tested at SPOJ
+ * edges such that the path from any leaf to the root contains at most log(n)
+ * light edges. Supports any segtree modifications/queries on paths and
+ * subtrees. Takes as input the full adjacency list.
+ * Status: Will do in morning
  */
 #pragma once
 
-#include "../data-structures/SegmentTree.h"
-
-typedef vector<pii> vpi;
-
-struct Node {
-	int d, par, val, chain = -1, pos = -1;
-};
-
-struct Chain {
-	int par, val;
-	vector<int> nodes;
-	Tree tree;
-};
-
-struct HLD {
-	typedef int T;
-	const T LOW = -(1<<29);
-	void f(T& a, T b) { a = max(a, b); }
-
-	vector<Node> V;
-	vector<Chain> C;
-
-	HLD(vector<vpi>& g) : V(sz(g)) {
-		dfs(0, -1, g, 0);
-		trav(c, C) {
-			c.tree = {sz(c.nodes), 0};
-			for (int ni : c.nodes)
-				c.tree.update(V[ni].pos, V[ni].val);
+#include "../data-structures/LazySegmentTree.h"
+
+template <bool USE_EDGES> struct HLD {
+	int N;
+	vector<vi> adj;
+	vi par, siz, depth, rt, pos;
+	Node *tree;
+	HLD(vector<vi> adj_)
+		: N(sz(adj_)), adj(adj_), par(N, -1), siz(N, 1), depth(N),
+		  rt(N), pos(N), tree(new Node(0, N)){ dfsSz(),dfsHld(); }
+	void dfsSz(int v = 0) {
+		if (par[v] != -1) adj[v].erase(find(all(adj[v]), par[v]));
+		trav(u, adj[v]) {
+			par[u] = v, depth[u] = depth[v] + 1;
+			dfsSz(u);
+			siz[v] += siz[u];
+			if (siz[u] > siz[adj[v][0]]) swap(u, adj[v][0]);
 		}
 	}
-
-	void update(int node, T val) {
-		Node& n = V[node]; n.val = val;
-		if (n.chain != -1) C[n.chain].tree.update(n.pos, val);
-	}
-
-	int pard(Node& nod) {
-		if (nod.par == -1) return -1;
-		return V[nod.chain == -1 ? nod.par : C[nod.chain].par].d;
+	int tim = 0;
+	void dfsHld(int v = 0) {
+		pos[v] = tim++;
+		trav(u, adj[v]) {
+			rt[u] = (u == adj[v][0] ? rt[v] : u);
+			dfsHld(u);
+		}
 	}
-
-	// query all *edges* between n1, n2
-	pair<T, int> query(int i1, int i2) {
-		T ans = LOW;
-		while(i1 != i2) {
-			Node n1 = V[i1], n2 = V[i2];
-			if (n1.chain != -1 && n1.chain == n2.chain) {
-				int lo = n1.pos, hi = n2.pos;
-				if (lo > hi) swap(lo, hi);
-				f(ans, C[n1.chain].tree.query(lo, hi));
-				i1 = i2 = C[n1.chain].nodes[hi];
-			} else {
-				if (pard(n1) < pard(n2))
-					n1 = n2, swap(i1, i2);
-				if (n1.chain == -1)
-					f(ans, n1.val), i1 = n1.par;
-				else {
-					Chain& c = C[n1.chain];
-					f(ans, n1.pos ? c.tree.query(n1.pos, sz(c.nodes))
-					              : c.tree.s[1]);
-					i1 = c.par;
-				}
-			}
+	template <class B> void process(int u, int v, B op) {
+		for (; rt[u] != rt[v]; v = par[rt[v]]) {
+			if (depth[rt[u]] > depth[rt[v]]) swap(u, v);
+			op(pos[rt[v]], pos[v] + 1);
```

Comment:
Not super happy that we assume commutativity here, but I guess fixing that takes up a bunch of space (and is also doable during contest)

Reply:
Resolved in another thread - supporting non-commutative functions requires substantially more code and is fairly rare.